mm2src/coins/utxo/utxo_common.rs

```diff
@@ -550,82 +540,146 @@ impl<'a, T: AsRef<UtxoCoinFields> + UtxoTxGenerationOps> UtxoTxBuilder<'a, T> {
-                    }
-                    self.sum_inputs >= outputs_plus_fee
-                } else {
-                    false
-                }
+                sum_output += self.total_tx_fee();
             },
-            FeePolicy::DeductFromOutput(_) => {
-                if self.sum_inputs >= self.sum_outputs_value {
-                    self.change = self.sum_inputs - self.sum_outputs_value;
-                    if self.change > self.dust() {
-                        if let ActualTxFee::Dynamic(ref f) = actual_tx_fee {
-                            self.tx_fee += (f * P2PKH_OUTPUT_LEN) / KILO_BYTE;
-                        }
-                    }
-                    if let Some(min_relay) = self.min_relay_fee {
-                        if self.tx_fee < min_relay {
-                            self.tx_fee = min_relay;
-                        }
+            FeePolicy::DeductFromOutput(_) => {},
+        };
+        sum_output
+    }
+
+    fn add_tx_inputs(&mut self, amount: u64) -> u64 {
+        self.tx.inputs.clear();
+        let mut total = 0u64;
+        for utxo in self.required_inputs.clone() {
+            self.tx.inputs.push(UnsignedTransactionInput {
+                previous_output: utxo.outpoint,
+                prev_script: utxo.script,
+                sequence: SEQUENCE_FINAL,
+                amount: utxo.value,
+            });
+            total += utxo.value;
+        }
+
+        for utxo in self.available_inputs.clone() {
+            if total >= amount {
+                break;
+            }
+            self.tx.inputs.push(UnsignedTransactionInput {
+                previous_output: utxo.outpoint,
+                prev_script: utxo.script,
+                sequence: SEQUENCE_FINAL,
+                amount: utxo.value,
+            });
+            total += utxo.value;
+        }
+        total
+    }
+
+    fn add_tx_outputs(&mut self) -> u64 {
+        self.tx.outputs.clear();
+        let mut total = 0u64;
+        for output in self.outputs.clone() {
+            total += output.value;
+            self.tx.outputs.push(output);
+        }
+        total
+    }
+
+    /// Adds change output.
+    /// Returns change value and dust change
+    fn add_change(&mut self, change_script_pubkey: &Bytes) -> (u64, u64) {
+        if self.sum_inputs > self.sum_outputs + self.total_tx_fee() {
+            let change = self.sum_inputs - (self.sum_outputs + self.total_tx_fee());
+            if change > self.dust() {
+                self.tx.outputs.push({
+                    TransactionOutput {
+                        value: change,
+                        script_pubkey: change_script_pubkey.clone(),
                     }
-                    true
-                } else {
-                    false
-                }
+                });
+                (change, 0u64)
+            } else {
+                (0u64, change)
+            }
+        } else {
+            (0u64, 0u64)
+        }
+    }
```

Comment:
what about ?
```rust
fn add_change(&mut self, change_script_pubkey: &Bytes) -> (u64, u64) {
        let sum_output_with_fee = self.sum_outputs + self.total_tx_fee();
        if self.sum_inputs > sum_output_with_fee {
            return (0u64, 0u64);
        }

        let change = self.sum_inputs - (sum_output_with_fee);
        if change > self.dust() {
            return (change, 0u64);
        };

        self.tx.outputs.push({
            TransactionOutput {
                value: change,
                script_pubkey: change_script_pubkey.clone(),
            }
        });

        (0u64, change)
    }
```

Reply:
refactored, thanks [40cfbda](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2316/commits/40cfbda989810aa5a0e83223fe7c1768cc2ed9b0).
there is also fix: ensure change >= dust (was: change > dust)

Reply:
Yeah, that's what too many nested code can do...:p